src/Open3D/Visualization/Rendering/Filament/FilamentScene.cpp

```diff
@@ -0,0 +1,697 @@
+                                  std::uint32_t w,
+                                  std::uint32_t h) {
+    auto handle = ViewHandle::Next();
+    auto view =
+            std::make_unique<FilamentView>(engine_, *this, resourceManager_);
+
+    view->SetViewport(x, y, w, h);
+    if (!views_.empty()) {
+        view->SetDiscardBuffers(View::TargetBuffers::DepthAndStencil);
+    }
+
+    ViewContainer c;
+    c.view = std::move(view);
+    views_.emplace(handle, std::move(c));
+
+    return handle;
+}
+
+View* FilamentScene::GetView(const ViewHandle& viewId) const {
+    auto found = views_.find(viewId);
+    if (found != views_.end()) {
+        return found->second.view.get();
+    }
+
+    return nullptr;
+}
+
+void FilamentScene::SetViewActive(const ViewHandle& viewId, bool isActive) {
+    auto found = views_.find(viewId);
+    if (found != views_.end()) {
+        found->second.isActive = isActive;
+    }
+}
+
+void FilamentScene::RemoveView(const ViewHandle& viewId) {
+    views_.erase(viewId);
+}
+
+GeometryHandle FilamentScene::AddGeometry(
+        const geometry::Geometry3D& geometry) {
+    GeometryHandle handle;
+
+    const auto geometryType = geometry.GetGeometryType();
+
+    MaterialInstanceHandle materialInstance;
+    handle = AddGeometry(geometry, materialInstance);
+
+    if (geometryType == geometry::Geometry::GeometryType::TriangleMesh) {
+        const auto& mesh = static_cast<const geometry::TriangleMesh&>(geometry);
+
+        if (mesh.HasMaterials()) {  // Mesh with materials
+            materialInstance = resourceManager_.CreateFromDescriptor(
+                    mesh.materials_.begin()->second);
+        } else if (mesh.HasVertexColors()) {  // Mesh with vertex color
+                                              // attribute set
+            materialInstance = resourceManager_.CreateMaterialInstance(
+                    defaults_mapping::ColorOnlyMesh);
+
+        } else if (mesh.HasTextures()) {  // Mesh with textures
+            materialInstance = resourceManager_.CreateMaterialInstance(
+                    defaults_mapping::Mesh);
+
+            auto wMaterial =
+                    resourceManager_.GetMaterialInstance(materialInstance);
+            auto mat = wMaterial.lock();
+
+            auto hTexture = resourceManager_.CreateTexture(
+                    mesh.textures_[0].FlipVertical());
+
+            if (hTexture) {
+                auto& entity = entities_[handle];
+                entity.texture = hTexture;
+
+                auto wTexture = resourceManager_.GetTexture(hTexture);
+                auto tex = wTexture.lock();
+                if (tex) {
+                    static const auto kDefaultSampler =
+                            FilamentMaterialModifier::
+                                    SamplerFromSamplerParameters(
+                                            TextureSamplerParameters::Pretty());
+                    mat->setParameter("texture", tex.get(), kDefaultSampler);
```

Comment:
I think this is too large of a task for now, given that I have never looked at this code before. To me it looks like this should always be a valid pointer, because the material instance being created is from kDefaultLit, which should always exist. If it doesn't exist, we will have aborted due to the error (or possibly either a string constant is wrong or our build didn't create the correct files). I've added an Asana task.